### ISSUE TITLE
Spike in browser's memory consumption in 2.0.32

### ISSUE THREAD

**Issue by Mintyboi** [emscripten-core/emscripten#16446]
I upgraded my emscripten toolchain from 2.0.26 -> 2.0.32 and found that there was a spike in the maximum memory consumption when loading my application.
It went from a peak of 1.2gb in 2.0.26 -> 3.2gb in 2.0.32 locally on my machine.

I don't have a reproducer just yet, but I thought I'd check if anyone knows anything about this or can list any suspicious areas where I can start investigating.

BTW, we reported an issue in Chrome on this same issue but there was not much movement there because it is not a regression in Chrome. https://bugs.chromium.org/p/chromium/issues/detail?id=1301979

**Comment by sbc100** [comment]
Could you share the configuration you are using for your build?   One easy way to do that is to list your entire link command (or at least all of the `-s` settings you are using when linking).

**Comment by Mintyboi** [comment]
@sbc100 Sure, here are my configuration.
I apologize for the long list. I have not yet narrowed down a smaller set
```-s DISABLE_EXCEPTION_CATCHING=0
-s MAIN_MODULE=2
-s INCLUDE_FULL_LIBRARY=1
-s AUTOLOAD_DYLIBS=0
-s NO_EXIT_RUNTIME=0
-s INLINING_LIMIT=1
-s FORCE_FILESYSTEM=1
--emit-symbol-map
-s LEGACY_GL_EMULATION=0
-s INITIAL_MEMORY=1000MB
-s ENVIRONMENT=web,worker
-s USE_SDL=0
-s EXTRA_EXPORTED_RUNTIME_METHODS=['allocate']
-s USE_WEBGL2=1
-s GL_TRACK_ERRORS=0
-s USE_GLFW=3
-s FULL_ES3=1
-s ALLOW_MEMORY_GROWTH=1
-s EVAL_CTORS=0
sideModule1.wasm
sideModule2.wasm
-s USE_ICU=1
-flto
-fPIC
-s WASM_BIGINT=1
-Os
```

**Comment by sbc100** [comment]
Do you have a build without dynamic linking?  I'd be curious to know if a build without `MAIN_MODULE` also saw a jump in usage. 

I don't have a lot of experience tracking down memory usage issues, but I'll see if I can find someone who does.

Can you confirm that the jump in memory is due to JS heap usage as opposed to linear wasm memory (i.e. does it happen even if you disallow ALLOW_MEMORY_GROWTH?)

**Comment by Mintyboi** [comment]
>Can you confirm that the jump in memory is due to JS heap usage as opposed to linear wasm memory (i.e. does it happen even if you disallow ALLOW_MEMORY_GROWTH?)

I don't think it's due to JS heap usage.
One clue is taking a look at the Javascript memory (live) in the Chrome's task manager and it doesn't spike fall when the Memory Footprint spikes or falls when Memory Footprint goes back to normal

**Comment by Mintyboi** [comment]
>Can you confirm that the jump in memory is due to JS heap usage as opposed to linear wasm memory (i.e. does it happen even if you disallow ALLOW_MEMORY_GROWTH?)

@sbc100 We tried these scenarios, however the memory spike still happens.
- disabling ALLOW_MEMORY_GROWTH
- reducing/removing the INITIAL_MEMORY
- removing `-s MAIN=2`

**Comment by sbc100** [comment]
Wow, that is interesting, I was pretty it was going to be one of those flags that was responsible.  

I wonder if we should be continuing our discussion in the chrome issue.  It should not matter that this is not a chrome regression (its clearly not), but it sure looks like a chrome bug.

Do you happen to have the both the old and new version of the app hosted somewhere so that we might look at the differences between them?

**Comment by dschuff** [comment]
Is the behavior similar on Chrome 99+? Chrome 99 introduces [dynamic tiering](https://v8.dev/blog/wasm-dynamic-tiering), which significantly changes how TurboFan compiles wasm modules on startup. Previously, Liftoff (baseline compiler) would compile the module as it downloaded, and then TF would start compiling the entire module as soon as it was instantiated. This meant that TF's compilation would saturate all CPUs and compete with the application for CPU time. With dynamic tiering enabled, TF only compiles functions after they become hot, meaning many functions get compiled later or not at all, which significantly improves CPU usage on startup. I *think* it launched in 99, but can you double-check using the flag, and see if it makes a difference?

**Comment by Mintyboi** [comment]
We tried out dynamic tiering when it was behind the feature flag and it alleviated this memory spike.
If dynamic tiering turned on by default in Chrome 99, I'll give v99 a try and see. Thanks for the info @dschuff.

Also just an FYI, we tried turning on wasm-exception-handling in our builds and that alleviated the memory spike as well.

**Comment by Mintyboi** [comment]
@dschuff I tried my application on ` Chrome 99.0.4844.51 (Official Build) (64-bit) (cohort: Stable)` but the issue seem to persist.
However the memory spike went away when I  launch chrome 99 with `--enable-blink-features=WebAssemblyDynamicTiering`.
As a matter of fact, it even improved what we previous had before this issue.
I'll continue to try it out on different machines to validate and update again.

**Comment by Mintyboi** [comment]
Update: We tried on several machines and it does not seem like dynamic tiering is enable by default in Chrome 99 yet.

**Comment by Mintyboi** [comment]
>Do you happen to have the both the old and new version of the app hosted somewhere so that we might look at the differences between them?

@sbc100 We are in discussions on how best to create a separate deployment and should be able to get some estimates next week. I'll keep you updated again.

**Comment by dschuff** [comment]
OK I found out that it's not just enabled by default in every M99, but it's rolling out gradually via a server-side flag. So it should be enabled everywhere on beta today, and is about to start rolling out a small percentage of stable clients starting in the next few days. That percentage will gradually ramp up, but will work on M99 and won't depend on users updating to M100.

**Comment by sbc100** [comment]
Do we have any idea why not enabling dynamic tiering would lead to such large memory usage spikes?

**Comment by dschuff** [comment]
I think it's just because V8 immediately starts up turbofan to compile all the functions in the module (using multiple threads), which just takes a lot of memory for a large application, and this runs concurrently with the application's own startup code.

**Comment by sbc100** [comment]
Hmm.. but why would that not effect the version of autocad built with older emscripten?   This problem only starts with the emscripten upgrade.

**Comment by Mintyboi** [comment]
@sbc100 That was our question as well. Also, enabling wasm exception-handling seems to lower the memory consumption as well. Another mystery

**Comment by kripken** [comment]
It may be a combination of factors. The new build may have more inlining, for example, which means more work for TurboFan, and that work may be avoided by dynamic tiering. The total memory used depends both on the toolchain and the VM, so certain combinations may end up causing noticeable spikes.

**Comment by Mintyboi** [comment]
Update: We should be able to get the secondary deployment up by the end of the week for comparison / investigation.
Are we still interested in taking a look into comparing what's the differences in compiling the 2 different wasm modules and is there anyone specific whom we should tag?

**Comment by Mintyboi** [comment]
@sbc100 Here we have 2 different autocad builds with different versions of the emscripten toolchain:
- emsdk 2.0.26 -> https://web.autocad.com/index.24755be6875dbc7a0a8431e2878e3bc4e20ab2ae.html
- emsdk 2.0.32 (with memory issue) -> https://web.autocad.com/index.17ec9613208bd86ec7faec0199fe1789166a7b65.html

Note: Please use the links to access the respective builds instead of pressing F5 to refresh as refreshing the page will redirect us back to the latest build.

**Comment by sbc100** [comment]
Thank you for preparing those two builds.  I can indeed access them.  Unfortunately I'm about to go on vacation until the 11th so somebody might be need to investigate.   I will also update the chromium bug in case somebody there finds these builds useful for tracking down what is going on.

**Comment by backes** [comment]
Just wanted to take a look, but it seems like both URLs send me to https://web.autocad.com. Were they removed in the meantime?
It would also help to just attach the problematic Wasm module here, and I can check if the issue is memory consumption during compilation in V8.

**Comment by Mintyboi** [comment]
@backes Thanks for trying it out.
That's fine, the URL will get redirected to https://web.autocad.com, however the resources that it loads will be the respective ones as listed. (Note: This is also the reason why we recommend to access the links directly instead of pressing F5 when refreshing the page)
>emsdk 2.0.26 -> https://web.autocad.com/index.24755be6875dbc7a0a8431e2878e3bc4e20ab2ae.html
emsdk 2.0.32 (with memory issue) -> https://web.autocad.com/index.17ec9613208bd86ec7faec0199fe1789166a7b65.html

Yes, I'm not sure if I did correctly, but the traces showed that it happened during Turbofan compilation. I did a short tracing using chrome://tracing and got these logs
[2.0.32-v8+memory.gz](https://github.com/emscripten-core/emscripten/files/8505177/2.0.32-v8%2Bmemory.gz)
